File: app/src/main/java/com/ladajules/notflix/data/repository/ProfileRepository.java
package com.ladajules.notflix.data.repository;

import android.util.Log;

import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.Query;
import com.ladajules.notflix.data.model.Profile;
import com.ladajules.notflix.data.remote.FirebaseManager;
import com.ladajules.notflix.utils.Constants;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class ProfileRepository {

    private final FirebaseFirestore firestore;
    private static final String TAG = "ProfileRepository";

    public ProfileRepository() {
        this.firestore = FirebaseManager.getFirestore();
    }

    public interface ProfileCallback<T> {
        void onResult(boolean success, T data, Exception e);
    }

    public void createProfile(Profile profile, ProfileCallback<String> callback) {
        Log.d(TAG, "Creating profile for user: " + profile.getUserId());

        String profileId = profile.getId();
        if (profileId == null || profileId.isEmpty()) {
            profileId = firestore.collection(Constants.PROFILES_COLLECTION).document().getId();
            profile.setId(profileId);
        }

        firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .set(profile.toMap())
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Log.d(TAG, "Profile created successfully with ID: " + profile.getId());
                        callback.onResult(true, profile.getId(), null);
                    } else {
                        Log.e(TAG, "Failed to create profile", task.getException());
                        callback.onResult(false, null, task.getException());
                    }
                });
    }

    public void getProfilesForUser(String userId, ProfileCallback<List<Profile>> callback) {
        Log.d(TAG, "Getting profiles for user: " + userId);

        firestore.collection(Constants.PROFILES_COLLECTION)
                .whereEqualTo("userId", userId)
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful() && task.getResult() != null) {
                        List<Profile> profiles = new ArrayList<>();
                        for (DocumentSnapshot document : task.getResult().getDocuments()) {
                            Map<String, Object> data = document.getData();
                            if (data != null) {
                                // Add ID to the map if it's not already there
                                data.put("id", document.getId());
                                Profile profile = Profile.fromMap(data);
                                if (profile != null) {
                                    profiles.add(profile);
                                }
                            }
                        }
                        // Sort by createdAt
                        Collections.sort(profiles, (p1, p2) -> Long.compare(p1.getCreatedAt(), p2.getCreatedAt()));
                        
                        Log.d(TAG, "Found " + profiles.size() + " profiles");
                        callback.onResult(true, profiles, null);
                    } else {
                        Log.e(TAG, "Failed to get profiles", task.getException());
                        callback.onResult(false, null, task.getException());
                    }
                });
    }

    public void getProfile(String profileId, ProfileCallback<Profile> callback) {
        Log.d(TAG, "Getting profile: " + profileId);

        firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful() && task.getResult() != null) {
                        DocumentSnapshot document = task.getResult();
                        if (document.exists()) {
                            Map<String, Object> data = document.getData();
                            if (data != null) {
                                data.put("id", document.getId());
                                Profile profile = Profile.fromMap(data);
                                callback.onResult(true, profile, null);
                            } else {
                                callback.onResult(true, null, null);
                            }
                        } else {
                            callback.onResult(true, null, null);
                        }
                    } else {
                        Log.e(TAG, "Failed to get profile", task.getException());
                        callback.onResult(false, null, task.getException());
                    }
                });
    }

    public void updateProfile(String profileId, Map<String, Object> updates, ProfileCallback<Void> callback) {
        Log.d(TAG, "Updating profile: " + profileId);

        firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .update(updates)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Log.d(TAG, "Profile updated successfully");
                        callback.onResult(true, null, null);
                    } else {
                        Log.e(TAG, "Failed to update profile", task.getException());
                        callback.onResult(false, null, task.getException());
                    }
                });
    }

    public void deleteProfile(String profileId, ProfileCallback<Void> callback) {
        Log.d(TAG, "Deleting profile: " + profileId);

        firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .delete()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Log.d(TAG, "Profile deleted successfully");
                        callback.onResult(true, null, null);
                    } else {
                        Log.e(TAG, "Failed to delete profile", task.getException());
                        callback.onResult(false, null, task.getException());
                    }
                });
    }

    public void getProfileCount(String userId, ProfileCallback<Integer> callback) {
        firestore.collection(Constants.PROFILES_COLLECTION)
                .whereEqualTo("userId", userId)
                .get()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful() && task.getResult() != null) {
                        callback.onResult(true, task.getResult().size(), null);
                    } else {
                        callback.onResult(false, 0, task.getException());
                    }
                });
    }
}

File: app/src/main/java/com/ladajules/notflix/data/repository/ProfileRepository.kt
package com.ladajules.notflix.data.repository

import android.util.Log
import com.google.firebase.firestore.FirebaseFirestore
import com.ladajules.notflix.data.model.Profile
import com.ladajules.notflix.data.remote.FirebaseManager
import com.ladajules.notflix.utils.Constants
import kotlinx.coroutines.tasks.await

class ProfileRepository {

    private val firestore: FirebaseFirestore = FirebaseManager.firestore
    private val TAG = "ProfileRepository"

    suspend fun createProfile(profile: Profile): Result<String> {
        return try {
            Log.d(TAG, "Creating profile for user: ${profile.userId}")
            
            // Generate ID if not provided
            val profileId = profile.id.ifEmpty { 
                firestore.collection(Constants.PROFILES_COLLECTION).document().id 
            }
            
            val profileWithId = profile.copy(id = profileId)
            
            firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .set(profileWithId.toMap())
                .await()
            
            Log.d(TAG, "Profile created successfully with ID: $profileId")
            Result.success(profileId)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create profile", e)
            Result.failure(e)
        }
    }

    suspend fun getProfilesForUser(userId: String): Result<List<Profile>> {
        return try {
            Log.d(TAG, "Getting profiles for user: $userId")
            
            val querySnapshot = firestore.collection(Constants.PROFILES_COLLECTION)
                .whereEqualTo("userId", userId)
                .get()
                .await()
            
            val profiles = querySnapshot.documents.mapNotNull { document ->
                document.data?.let { data ->
                    Profile.fromMap(data + ("id" to document.id))
                }
            }.sortedBy { it.createdAt }
            
            Log.d(TAG, "Found ${profiles.size} profiles")
            Result.success(profiles)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get profiles", e)
            Result.failure(e)
        }
    }

    suspend fun getProfile(profileId: String): Result<Profile?> {
        return try {
            Log.d(TAG, "Getting profile: $profileId")
            
            val document = firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .get()
                .await()
            
            val profile = if (document.exists()) {
                document.data?.let { data ->
                    Profile.fromMap(data + ("id" to document.id))
                }
            } else null
            
            Log.d(TAG, "Profile found: ${profile != null}")
            Result.success(profile)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get profile", e)
            Result.failure(e)
        }
    }

    suspend fun updateProfile(profileId: String, updates: Map<String, Any>): Result<Unit> {
        return try {
            Log.d(TAG, "Updating profile: $profileId")
            
            firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .update(updates)
                .await()
            
            Log.d(TAG, "Profile updated successfully")
            Result.success(Unit)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to update profile", e)
            Result.failure(e)
        }
    }

    suspend fun deleteProfile(profileId: String): Result<Unit> {
        return try {
            Log.d(TAG, "Deleting profile: $profileId")
            
            firestore.collection(Constants.PROFILES_COLLECTION)
                .document(profileId)
                .delete()
                .await()
            
            Log.d(TAG, "Profile deleted successfully")
            Result.success(Unit)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to delete profile", e)
            Result.failure(e)
        }
    }

    suspend fun getProfileCount(userId: String): Result<Int> {
        return try {
            val querySnapshot = firestore.collection(Constants.PROFILES_COLLECTION)
                .whereEqualTo("userId", userId)
                .get()
                .await()
            
            Result.success(querySnapshot.size())
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}